docs/spec/lightclient/detection.md

```diff
@@ -0,0 +1,491 @@
+> By verifying we mean calling `VerifyToTarget` from the
+> [[verification]] specification.
+
+## Definitions
+
+- A fixed set of full nodes is provided in the configuration upon
+     initialization. Initially this set is partitioned into
+    -  one full node that is the *primary* (singleton set),
+	-  a set *Secondaries* (of fixed size, e.g., 3),
+	-  a set *FullNodes*.
+- A set *FaultyNodes* of nodes that the light client suspects of being faulty; it is initially empty
+
+- *Lightstore* as defined in the [verification specification][verification].
+
+
+
+
+#### **[LCD-INV-NODES]:**
+The detector shall maintain the following invariants:
+   - *FullNodes \intersect Secondaries = {}*
+   - *FullNodes \intersect FaultyNodes = {}*
+   - *Secondaries \intersect FaultyNodes = {}*
+   
+and the following transition invariant
+   - *FullNodes' \union Secondaries' \union FaultyNodes' = FullNodes \union Secondaries \union FaultyNodes*
+
+## Solution
+
+
+
+### Inter Process Communication
+
+
+```go
+func FetchLightBlock(peer PeerID, height Height) LightBlock
+```
+See the [verification specification][verification] for details.
+
+### Auxiliary Functions (Local)
+
+
+```go
+Replace_Secondary(addr Address)
+```
+- Expected precondition
+    - *FullNodes* is nonempty
+- Expected postcondition
+    - addr is moved from *Secondaries* to *FaultyNodes*
+    - an address *a* is moved from *FullNodes* to *Secondaries*
+- Error condition
+    - if precondition is violated
+
+
+
+### From the verifier
+```go
+func VerifyToTarget(primary PeerID, lightStore LightStore,
+                    targetHeight Height) (LightStore, Result)
+```
+See the [verification specification][verification] for details.
+
+
+
+## Solution
+
+Shared data of the light client
+- a pool of full nodes *FullNodes* that have not been contacted before
+- peer set called *Secondaries*
+- primary
+- LightStore
+
+
+
+
+The problem is solved by calling  the function `ForkDetector` with
+a lightstore that contains a light block that has
+just been verified by the verifier. 
+
+
+```go
+func ForkDetector(ls LightStore)  {
+    Forks.Init // initialize a container in which we collect forks
+	for i, s range Secondaries {
+		sh := FetchLightBlock(s,LightStore.LatestVerified().Height)
+		// as the check below only needs the header, it is sufficient
+		// to download the header rather than the LighBlock
+		if LightStore.LatestVerified().Header == sh.Header {
+				// header matches. we do nothing.
+	    }
+		else {
+			    // [LCD-REQ-REP]
+			    // header does not match. there is a situation.
+				// we try to verify sh by querying s
+				// we set up an auxiliary lightstore with the highest
+			    // trusted lightblock and the lightblock we want to verify
+				auxLS.Init
+				auxLS.Update(LightStore.LatestTrusted(), StateVerified);
+				auxLS.Update(sh,StateUnverified);
+				result := VerifyToTarget(s, auxLS, LightStore.LatestVerified().Height)
+				if result = (_,ResultSuccess) || (_,EXPIRED) {
```

Comment:
We should clarify what is the expectation of VerifyToTarget so if it returns TimeoutError it can be assumed faulty. I guess that VerifyToTarget with correct full node should never terminate with TimeoutError.

Reply:
will do